### PR TITLE
Add TraceProcessor and move some existing functionality to rules

### PR DESCRIPTION
**Pull request by tylerbenson**
Specifically, the Status5XXRule, URLAsResourceNameRule, and Status404Rule.

I tried moving some of the service name rules, but discovered that service name must be set eagerly because of the inheritance of name by child spans.

### CODE REVIEW

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java [hunk 1]
```diff
@@ -86,11 +87,11 @@ private String normalizePath(final String path) {
...
   }
 
-  private String addMethodIfAvailable(final DDSpanContext context, String path) {
+  private String addMethodIfAvailable(final Map<String, String> meta, String path) {
     // if the verb (GET, POST ...) is present, add it
-    final String verb = (String) context.getTags().get(Tags.HTTP_METHOD.getKey());
+    final String verb = meta.get(Tags.HTTP_METHOD.getKey());
     if (verb != null && !verb.isEmpty()) {
-      path = verb + " " + path;
+      path = verb.toUpperCase() + " " + path;
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java]: Note, minor risk of a breaking change here in exchange for better consistency.

**Comment by dougqh** [comment]
It would help to have a high-level explanation of the new pieces concepts in the PR summary.

File: dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/TracingRequestHandler.java [hunk 2]
```diff
@@ -29,7 +29,7 @@ public AmazonWebServiceRequest beforeMarshalling(final AmazonWebServiceRequest r
 
   @Override
   public void beforeRequest(final Request<?> request) {
-    final AgentSpan span = startSpan("aws.command");
```
Comment (dougqh) [inline dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/TracingRequestHandler.java]: This change seems unrelated; why are we making it?

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import datadog.trace.common.processor.rule.URLAsResourceNameRule;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class TraceProcessor {
+  final Rule[] DEFAULT_RULES =
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java]: I'm debating whether placing the list of Rules in the TraceProcessor is the right choice.

An alternative would be to associate a set of rules with a span as construction, but to defer execution until the end.
I'm not sure if that's better or not in this case -- just suggesting an alternative.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      new Rule[] {
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java]: The List<Rule> + the loop in onTraceComplete effectively comprise a CompositeRule.
I think it might be better to explicit introduce a CompositeRule class.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  @Override
+  public String[] aliases() {
+    return new String[] {"DBStatementAsResourceName"};
+  }
+
+  @Override
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    if (tags.containsKey(Tags.DB_STATEMENT.getKey())) {
+      // Special case: Mongo
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java]: Mongo special case here for Mongo strikes me as problematic from both an encapsulation & maintenance standpoint.  This really makes me think that we might be better off attaching rules as each span construction -- or at least having an obvious way to override with a integration specific substitute rule.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+public class ErrorRule implements TraceProcessor.Rule {
+  @Override
+  public String[] aliases() {
+    return new String[] {"ErrorFlag"};
+  }
+
+  @Override
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    if (tags.containsKey(Tags.ERROR.getKey())) {
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java]: The instanceof checks here strike me as a bit messy as well.
I've wondered for awhile whether we should have a Tags class rather than using Map<String, Object>.
The error prone checks here -- lead to think the answer is yes.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/SpanTypeRule.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+public class SpanTypeRule implements TraceProcessor.Rule {
+  @Override
+  public String[] aliases() {
+    return new String[] {"SpanTypeDecorator"};
+  }
+
+  @Override
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    if (tags.containsKey(DDTags.SPAN_TYPE)) {
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/SpanTypeRule.java]: containsKey followed by get is redundant -- unless we're expecting maps with null values.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/SpanTypeRule.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      span.setSpanType(tags.get(DDTags.SPAN_TYPE).toString());
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/SpanTypeRule.java]: I think I understand what is happening here, but this looks odd and error prone.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/Status5XXRule.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    return new String[] {"Status5XXDecorator"};
+  }
+
+  @Override
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    if (!span.context().getErrorFlag() && tags.containsKey(Tags.HTTP_STATUS.getKey())) {
+      final Object value = tags.get(Tags.HTTP_STATUS.getKey());
+      final int responseCode =
+          value instanceof Integer ? (int) value : Integer.parseInt(value.toString());
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/Status5XXRule.java]: Not handling a NumberFormatException from parseInt.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java [hunk 10]
```diff
@@ -1,38 +1,40 @@
...
-    // do nothing if the status code is already set and equals to 404.
-    // TODO: it assumes that Status404Decorator is active. If it's not, it will lead to unexpected
-    // behaviors
-    if (value == null || statusCode != null && statusCode.equals("404")) {
-      return true;
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    final DDSpanContext context = span.context();
+    final Object httpStatus = tags.get(Tags.HTTP_STATUS.getKey());
+    if (context.isResourceNameSet()
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java]: This logic was repeated elsewhere.  I think we should find a better to encapsulate it.

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java [hunk 11]
```diff
@@ -10,15 +11,17 @@
...
 
   public List<DDSpan> firstTrace() {
     return get(0);
   }
 
   @Override
-  public void write(final List<DDSpan> trace) {
+  public void write(List<DDSpan> trace) {
     synchronized (latches) {
+      trace = processor.onTraceComplete(trace);
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java]: I don't think this needs to be under the lock.  Any synchronization for each span should be handled elsewhere.

**Comment by dougqh** [comment]
> I tried moving some of the service name rules, but discovered that service name must be set eagerly because of the inheritance of name by child spans.

This strikes me as a serious design issue that we need to revisit, since services are one of things that we commonly change via Decorators.

File: dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/TracingRequestHandler.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/TracingRequestHandler.java:32]: This is related to the removal of `OperationDecorator`.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java (hunk 3 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java:21]: I see how that could work, but not sure if it would offer significant benefit.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java (hunk 5 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java:23]: This was a carry over from `DBStatementAsResourceName` that I didn't want to unwind.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java:19]: I'm not opposed to this, but prob not in this PR.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java (hunk 10 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java:27]: This logic is similar, but not the same as `Status404Rule`

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java (hunk 11 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java:24]: Are you suggesting moving it outside the lock? Does it make much difference either way?

File: dd-trace-ot/src/test/groovy/datadog/opentracing/decorators/SpanDecoratorTest.groovy [hunk 12]
```diff
@@ -286,28 +276,29 @@ class SpanDecoratorTest extends DDSpecification {
     // these 2 lines to avoid checking for {@code DDSpanContext.DD_MEASURED} for every metric:
     span.context().setMetric(DDSpanContext.DD_MEASURED, 42)
     span.metrics.remove(DDSpanContext.DD_MEASURED, 42)
```
Comment (devinsba) [inline dd-trace-ot/src/test/groovy/datadog/opentracing/decorators/SpanDecoratorTest.groovy:278]: Looks like you can remove these two lines now that you aren't doing a equals against the whole object

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/Status5XXRule.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Override
+  public void processSpan(
+      final DDSpan span, final Map<String, Object> tags, final Collection<DDSpan> trace) {
+    if (!span.context().getErrorFlag() && tags.containsKey(Tags.HTTP_STATUS.getKey())) {
+      final Object value = tags.get(Tags.HTTP_STATUS.getKey());
+      try {
+        final int responseCode =
+            value instanceof Integer ? (int) value : Integer.parseInt(value.toString());
+        span.setError(500 <= responseCode && responseCode < 600);
```
Comment (randomanderson) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/Status5XXRule.java:24]: I know this is just a copy of `Status5XXDecorator` but how does this not break overriding the error range with `dd.http.server.error.statuses`?

Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/Status5XXRule.java:24]: Good question... I'll have to investigate that further.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java (hunk 5 above)
Comment (dougqh) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java:23]: Understandable, but if we're going to undertake a new design, I'd like the new design to have the ability to unwind that mess.  Right now, it isn't clear to me that there's a way to do so.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+      // Skip the decorators
+      if (tags.containsKey(Tags.COMPONENT.getKey())
+          && "java-mongo".equals(tags.get(Tags.COMPONENT.getKey()))) {
+        return;
+      }
+
+      final String statement = tags.get(Tags.DB_STATEMENT.getKey()).toString();
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/DBStatementRule.java:30]: Doesn't this have the potential to raise an NPE?

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java (hunk 6 above)
Comment (dougqh) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java:19]: Okay

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java (hunk 10 above)
Comment (dougqh) [reply dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java:27]: Yes, there are slight differences, but I think some sharing is possible.

isResourceNameSet() and statusCode checks are the same.
However if there was a helper class to handle the int conversion and comparisons, I think the repetition would be smaller and less problematic.

I'm okay with not fixing this now, but I think there is a problem here.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java [hunk 15]
```diff
@@ -86,11 +88,11 @@ private String normalizePath(final String path) {
...
   }
 
-  private String addMethodIfAvailable(final DDSpanContext context, String path) {
+  private String addMethodIfAvailable(final Map<String, Object> meta, String path) {
     // if the verb (GET, POST ...) is present, add it
-    final String verb = (String) context.getTags().get(Tags.HTTP_METHOD.getKey());
-    if (verb != null && !verb.isEmpty()) {
-      path = verb + " " + path;
+    final Object verb = meta.get(Tags.HTTP_METHOD.getKey());
+    if (verb != null && !verb.toString().isEmpty()) {
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/URLAsResourceNameRule.java:94]: This looks better than the prior version, but it would be nice to eliminate the second call to verb.toString()

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java (hunk 11 above)
Comment (dougqh) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java:24]: Right, this shouldn't be under the lock.

The extra work will increase the lock hold time -- and that can actually have quite the impact.
For instance, adaptive spinning could cause the lock to be promoted from a spin lock to full lock.

But more importantly, the synchronization here needs to come from the spans not this lock.
Any additional synchronization isn't helping and just increases the likelihood of a deadlock.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      if (isEnabled(rule)) {
+        rules.add(rule);
+      }
+    }
+  }
+
+  private static boolean isEnabled(final Rule rule) {
+    boolean enabled = Config.get().isDecoratorEnabled(rule.getClass().getSimpleName());
+    for (final String alias : rule.aliases()) {
+      enabled &= Config.get().isDecoratorEnabled(alias);
```
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/TraceProcessor.java:49]: Are you still calling them decorators?
I understanding supporting the old config, but maybe the method name should be changed.

File: dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java (hunk 6 above)
Comment (dougqh) [inline dd-trace-ot/src/main/java/datadog/trace/common/processor/rule/ErrorRule.java:19]: containsKey / get again -- strictly speaking, this probably isn't thread safe.

File: dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java (hunk 11 above)
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java:24]: `ListWriter` is primarily used for unit tests.  I'm not so sure it will make a big difference either way.  I can move it outside though if you really want me to.

Comment (dougqh) [reply dd-trace-ot/src/main/java/datadog/trace/common/writer/ListWriter.java:24]: Fair enough.

We've got a bit of problem using very coarse grained locking -- even around concurrent collections, so this jumped out at me, but I suppose in this instance, it is okay.

**Comment by dougqh** [comment]
Given some of the shortcomings, I fully expect we'll need to revisit this design at some point.
But this is also an important step forward in terms of fixing some resource naming issues, so I think it is a solid improvement as is.
